modules/hostmatcher/hostmatcher.go

```diff
@@ -43,52 +47,118 @@ func ParseHostMatchList(hostList string) *HostMatchList {
 		if err == nil {
 			hl.ipNets = append(hl.ipNets, ipNet)
 		} else {
-			hl.hosts = append(hl.hosts, s)
+			hl.patterns = append(hl.patterns, s)
 		}
 	}
 	return hl
 }
 
-// MatchesHostOrIP checks if the host or IP matches an allow/deny(block) list
-func (hl *HostMatchList) MatchesHostOrIP(host string, ip net.IP) bool {
-	var matched bool
-	host = strings.ToLower(host)
-	ipStr := ip.String()
-loop:
-	for _, hostInList := range hl.hosts {
-		switch hostInList {
+// ParseSimpleMatchList parse a simple matchlist (no built-in networks, no CIDR support)
+func ParseSimpleMatchList(settingKeyHint string, matchList string) *HostMatchList {
+	hl := &HostMatchList{
+		SettingKeyHint: settingKeyHint,
+		SettingValue:   matchList,
+	}
+	for _, s := range strings.Split(matchList, ",") {
+		s = strings.ToLower(strings.TrimSpace(s))
+		if s == "" {
+			continue
+		}
+		if s == MatchBuiltinLoopback || s == MatchBuiltinPrivate || s == MatchBuiltinExternal {
+			// for built-in patterns, we convert it from "private" => "[p]rivate" for internal usage and keep the same result as `matchlist`
+			hl.patterns = append(hl.patterns, "["+s[:1]+"]"+s[1:])
```

Comment:
This seems like a hack and a fragile one too.

Shouldn't we be using an interface for these patterns, e.g.:

```go
type HostMatcher interface {
   MatchIP(ip net.IP) bool
   MatchHost(hostname string) bool
}
```

```go
type AllHostMatcher struct {}

func (*AllHostMatcher) MatchIP(_ net.IP) bool { return true; }

func (*AllHostMatcher) MatchHostname(_ string) bool { return true; }
```

Reply:
I haven't got a better idea about this. Since we need to do mixed checking with host/ip/wildcard, the logic may seem fragile.

Do you have a better design? Or you can use this PR to rework. As long as the problem can be resolve, all are fine to me.

Reply:
Maybe I can the struct like this:

```
type HostMatchList struct {
  builtin   []string
	patterns []string
	ipNets   []*net.IPNet
}
```

Then the logic can be more clear. Will push a commit.

Reply:
Reworked. Please review again